src/data/logs.ts

```diff
@@ -0,0 +1,264 @@
+        totalLength
+      );
+    }
+    for (let pointIndex = 0; pointIndex < totalLength; pointIndex++) {
+      const currentValue = df.frame.get(pointIndex).Value;
+      const valueToAdd = field.values.get(pointIndex);
+      const totalValue = currentValue === null && valueToAdd === null ? null : (currentValue || 0) + (valueToAdd || 0);
+      if (totalValue > 0 && !df.hasNonZeroValues) {
+        df.hasNonZeroValues = true;
+      }
+      df.frame.set(pointIndex, { Value: totalValue, Time: timeField.values.get(pointIndex) });
+    }
+    logLevelToDataFrame.set(logLevel, df);
+  });
+
+  // if we have all zeroes for a particular level, we want to exclude it from the logs volume histogram
+  return [...logLevelToDataFrame.values()].reduce((acc: MutableDataFrame[], { frame, hasNonZeroValues }) => {
+    if (hasNonZeroValues) {
+      acc.push(frame);
+    }
+    return acc;
+  }, []);
+}
+
+/**
+ * Returns field configuration used to render logs volume bars
+ */
+function getLogVolumeFieldConfig(level: LogLevel, oneLevelDetected: boolean) {
+  const name = oneLevelDetected && level === LogLevel.unknown ? 'logs' : level;
+  const color = LogLevelColor[level];
+  return {
+    displayNameFromDS: name,
+    color: {
+      mode: FieldColorModeId.Fixed,
+      fixedColor: color,
+    },
+    custom: {
+      drawStyle: GraphDrawStyle.Bars,
+      barAlignment: BarAlignment.Center,
+      lineColor: color,
+      pointColor: color,
+      fillColor: color,
+      lineWidth: 1,
+      fillOpacity: 100,
+      stacking: {
+        mode: StackingMode.Normal,
+        group: 'A',
+      },
+    },
+  };
+}
+
+export function getIntervalInfo(scopedVars: ScopedVars, timespanMs: number): { interval: string; intervalMs?: number } {
+  if (scopedVars.__interval) {
+    let intervalMs: number = scopedVars.__interval_ms.value;
+    let interval;
+    // below 5 seconds we force the resolution to be per 1ms as interval in scopedVars is not less than 10ms
+    if (timespanMs < SECOND * 5) {
+      intervalMs = MILLISECOND;
+      interval = '1ms';
+    } else if (intervalMs > HOUR) {
+      intervalMs = DAY;
+      interval = '1d';
+    } else if (intervalMs > MINUTE) {
+      intervalMs = HOUR;
+      interval = '1h';
+    } else if (intervalMs > SECOND) {
+      intervalMs = MINUTE;
+      interval = '1m';
+    } else {
+      intervalMs = SECOND;
+      interval = '1s';
+    }
+
+    return { interval, intervalMs };
+  } else {
+    return { interval: '$__interval' };
+  }
+}
+
+export function getTimeFieldRoundingClause(
+  scopedVars: ScopedVars,
+  timespanMs: number,
+  timeField = 'created_at'
+): string {
+  let interval = 'DAY';
+  if (scopedVars.__interval) {
+    let intervalMs: number = scopedVars.__interval_ms.value;
+    // below 5 seconds we force the resolution to be per 1ms as interval in scopedVars is not less than 10ms
+    if (timespanMs < SECOND * 5) {
+      // TODO: workaround
```

Comment:
Is it necessary to support millisecond precision, or can we get away with up to second precision?

Reply:
I am excluding this for now.